### PR TITLE
Add monte-carlo solver

### PR DESCRIPTION
**Pull request by AlexLipp**
Add method to SampleNetwork class that performs a monte carlo estimation of uncertainty on predictions (upstream and downstream) by propagating a provided relative observational error through the unmixing routine.

### CODE REVIEW

File: geochem_inverse_optimize.py [hunk 1]
```diff
@@ -202,6 +208,38 @@ def solve(
...
+                (sample, value * np.random.normal(loc=1, scale=relative_error / 100))
+                for sample, value in observation_data_resampled.items()
+            )
+            element_pred_down, element_pred_upstream = self.solve(
+                observation_data=observation_data_resampled,
+                solver=solver,
+                regularization_strength=regularization_strength,
+            )  # Solve problem
+            for sample_name in element_pred_down.keys():
+                if i == 1:  # Initiate output array
```
Comment (r-barnes) [inline geochem_inverse_optimize.py:234]: Use `defaultdict` to simplify the logic here.

File: unmix_mwe.py [hunk 2]
```diff
@@ -15,7 +17,7 @@
 obs_data = pd.read_csv("data/geochem_no_dupes.dat", delimiter=" ")
 obs_data = obs_data.drop(columns=["Bi", "S"])
 
-element = "Mg"  # Set element
+element = "Be"  # Set element
```
Comment (r-barnes) [inline unmix_mwe.py:20]: Why is this change in this PR?

File: unmix_mwe.py [hunk 3]
```diff
@@ -27,17 +29,51 @@
 element_data = gio.get_element_obs(
     element, obs_data
 )  # Return dictionary of {sample_name:concentration}
+
+regularizer_strength = 10 ** (-2.6)
```
Comment (r-barnes) [inline unmix_mwe.py:33]: It's probably better to put this near the top of the script as
```
regularizer_strength: Final[float] = 10 ** (-2.6)
```

File: unmix_mwe.py [hunk 4]
```diff
@@ -27,17 +29,51 @@
...
 gio.visualise_downstream(pred_dict=element_pred_down, obs_dict=element_data, element=element)
+
+# Add uncertainties from monte-carlo sampling
+obs = []
+pred = []
+pred_err = []
+for sample in element_data.keys():
+    obs += [element_data[sample]]
+    pred += [element_pred_down[sample]]
+    pred_err += [downstream_uncerts[sample]]
```
Comment (r-barnes) [inline unmix_mwe.py:71]: Use a list comprehension to simplify this.

**Review by r-barnes — CHANGES_REQUESTED** [review]

File: geochem_inverse_optimize.py [hunk 5]
```diff
@@ -202,6 +208,38 @@ def solve(
...
+        observation_data: ElementData,
+        relative_error: float,
+        num_repeats: int,
+        regularization_strength: float,
+        solver: str = "gurobi",
+    ):
+        predictions_down_mc = {}
+        predictions_up_mc = {}
+        i = 1
+        while i < num_repeats:
```
[the remark this answers is not in the document]
Comment (AlexLipp) [reply geochem_inverse_optimize.py:222]: Agreed. Added in for loop.

File: geochem_inverse_optimize.py (hunk 1 above)
Comment (AlexLipp) [reply geochem_inverse_optimize.py:234]: Great tip thanks!

File: unmix_mwe.py [hunk 6]
```diff
@@ -1,8 +1,11 @@
 # Preamble
 import os
 import sys
+from platform import release
```
Comment (r-barnes) [inline unmix_mwe.py:4]: This looks unused?

Comment (AlexLipp) [reply unmix_mwe.py:4]: Removed

File: geochem_inverse_optimize.py [hunk 7]
```diff
@@ -25,7 +26,9 @@ def cp_log_ratio_norm(a, b):
     return cp.maximum(a / b, b * cp.inv_pos(a))
 
 
-def nx_topological_sort_with_data(G: nx.DiGraph) -> Iterator[Tuple[str, pyfastunmix.SampleNode]]:
+def nx_topological_sort_with_data(
```
[the remark this answers is not in the document]
Comment (AlexLipp) [reply geochem_inverse_optimize.py:29]: i checked the pre-commit .yaml which has line-length at 100... I've also checked that this is the line-length being used for black in my VSCODE too.

File: geochem_inverse_optimize.py [hunk 8]
```diff
@@ -198,13 +199,38 @@ def solve(
...
+            }
+            element_pred_down, element_pred_upstream = self.solve(
+                observation_data=observation_data_resampled,
+                solver=solver,
+                regularization_strength=regularization_strength,
+            )  # Solve problem
+            for sample_name in element_pred_down.keys():
+                predictions_down_mc[sample_name] += [element_pred_down[sample_name]]
+                predictions_up_mc[sample_name] += [element_pred_upstream[sample_name]]
+            i += 1
```
Comment (r-barnes) [inline geochem_inverse_optimize.py:225]: This seems incorrect.

**Review by r-barnes — CHANGES_REQUESTED** [review]
I've done some rebasing to handle the "Squash + Rebase" workflow I used on the string-graph refactor. Could you make sure this PR is still good?

File: geochem_inverse_optimize.py (hunk 8 above)
Comment (AlexLipp) [reply geochem_inverse_optimize.py]: The `i+=1' isn't need so I removed it. 
I checked that this still produces the expected outcomes and it does.

**Comment by AlexLipp** [comment]
I've checked that this still works as expected after the squash+rebase. An import declaration had been removed but has been readded.

**Comment by r-barnes** [comment]
@AlexLipp : I'm still seeing a conflict with unmix_mwe.py, though.

**Comment by AlexLipp** [comment]
Oh yes sorry if I wasn’t clear. I’ve inspected this conflict directly and it’s trivial. I was going to do it myself but left it for you just in case you had some protocol on conflict resolution and merging. It’s simply the ordering of two commands which is throwing everything out of whack but has not got major consequences/conflicts.

On 18 Nov 2022, at 21:57, Richard Barnes ***@***.***> wrote:

﻿

@AlexLipp<https://github.com/AlexLipp> : I'm still seeing a conflict with unmix_mwe.py, though.

**Pull request merged**